cg/meta/deliver/delivery_api.py

```diff
@@ -0,0 +1,245 @@
+            self.deliver_files(case=case, pipeline=pipeline)
+
+    def _deliver_files_by_case(self, case_id: str, pipeline: str) -> None:
+        case: Case | None = self.store.get_case_by_internal_id(case_id)
+        if not case:
+            LOG.warning(f"Could not find case {case_id}")
+            return
+        self.deliver_files(case=case, pipeline=pipeline)
+
+    def deliver_files(self, case: Case, pipeline: str):
+        if not self._case_is_deliverable(case=case, pipeline=pipeline):
+            return
+        self._deliver_case_files(case=case, pipeline=pipeline)
+        self._deliver_sample_files(case=case, pipeline=pipeline)
+
+    def _deliver_case_files(self, case: Case, pipeline: str) -> None:
+        LOG.debug(f"Deliver case files for {case.internal_id}")
+
+        if not get_case_tags_for_pipeline(pipeline):
+            return
+
+        version: Version = self.hk_api.last_version(case.internal_id)
+        out_dir: Path = self._create_delivery_directory(case)
+        self._link_case_files(case=case, version=version, pipeline=pipeline, out_dir=out_dir)
+
+    def _link_case_files(self, case: Case, version: Version, pipeline: str, out_dir: Path):
+        sample_ids: set[str] = self._get_sample_ids_for_case(case)
+        files: Iterable[Path] = self._get_case_files_from_version(
+            version=version, sample_ids=sample_ids, pipeline=pipeline
+        )
+        for file_path in files:
+            out_path: Path = get_out_path(out_dir=out_dir, file=file_path, case=case)
+            create_link(source=file_path, destination=out_path, dry_run=self.dry_run)
+
+    def _get_sample_ids_for_case(self, case: Case) -> set[str]:
+        links = self.store.get_case_samples_by_case_id(case.internal_id)
+        samples: list[Sample] = [link.sample for link in links]
+        sample_ids: set[str] = {sample.internal_id for sample in samples}
+        return sample_ids
+
+    def _create_delivery_directory(self, case: Case) -> Path:
+        delivery_base = get_delivery_dir_path(
+            case_name=case.name,
+            customer_id=case.customer.internal_id,
+            ticket=case.latest_ticket,
+            base_path=self.project_base_path,
+        )
+        if not self.dry_run:
+            LOG.debug(f"Creating project path {delivery_base}")
+            delivery_base.mkdir(parents=True, exist_ok=True)
+        return delivery_base
+
+    def _deliver_sample_files(self, case: Case, pipeline: str) -> None:
+        if not get_sample_tags_for_pipeline(pipeline):
+            return
+
+        deliverable_samples = self._get_deliverable_samples(case=case, pipeline=pipeline)
+        case_name: str | None = get_delivery_case_name(case=case, pipeline=pipeline)
+        for link in deliverable_samples:
+            delivery_base: Path = self._create_sample_delivery_directory(
+                case=case, sample=link.sample, pipeline=pipeline
+            )
+
+            number_linked_files: int = 0
```

Comment:
That would change the behavior of the code. 
Instead of processing other samples first, it will raise an exception as soon as a sample without files is found.
I'll try to clean it up